### ISSUE TITLE
`ExposurePipeline` uses `meta.filename` instead of actual filename

### ISSUE THREAD

**Issue by braingram** [spacetelescope/romancal#818]
I believe the changes in #802 are not compatible with the `test_processing_pipeline_all_saturated` regtest.

Here's a recent regtest run with this PR where only the mentioned test was run and stdout was not captured:
https://plwishmaster.stsci.edu:8081/blue/organizations/jenkins/RT%2FRoman-Developers-Pull-Requests/detail/Roman-Developers-Pull-Requests/311/pipeline/205/#step-206-log-97

A few relevant log lines are:
```
2023-08-08 16:45:59,578 - stpipe.ExposurePipeline - INFO - Starting Roman exposure calibration pipeline ...
2023-08-08 16:45:59,818 - stpipe.ExposurePipeline - INFO - Processing a WFI exposure <roman_datamodels.datamodels._datamodels.ScienceRawModel object at 0x7f0893db6450>
...
2023-08-08 16:51:22,126 - stpipe.ExposurePipeline - INFO - Roman exposure calibration pipeline ending...

2023-08-08 16:51:24,208 - stpipe.ExposurePipeline - INFO - Saved model in r0000101001001001001_01101_0001_WFI01_cal.asdf
```
Note the filename above does not match the output filename in the regression test:
https://github.com/spacetelescope/romancal/blob/043776069a21ee6f92f6af356293e4f4bec202bb/romancal/regtest/test_wfi_pipeline.py#L438
which causes the error
```
E                   FileNotFoundError: [Errno 2] No such file or directory: '/srv/jenkins/workspace/RT/Roman-Developers-Pull-Requests/clone/test_outputs/popen-gw1/test_processing_pipeline_all_s0/r0000101001001001001_01101_0001_WFI01_ALL_SATURATED_cal.asdf'
/srv/jenkins/workspace/RT/Roman-Developers-Pull-Requests/miniconda/envs/tmp_env0/lib/python3.11/site-packages/asdf/generic_io.py:1150: FileNotFoundError
```

This seems related to the input filename handling in exposure pipline where when provided with an `input` as a string, `input` is overwritten as a datamodel:
https://github.com/spacetelescope/romancal/blob/043776069a21ee6f92f6af356293e4f4bec202bb/romancal/pipeline/exposure_pipeline.py#L78-L85
then added to `expos_file`:
https://github.com/spacetelescope/romancal/blob/043776069a21ee6f92f6af356293e4f4bec202bb/romancal/pipeline/exposure_pipeline.py#L94-L101
before this check:
https://github.com/spacetelescope/romancal/blob/043776069a21ee6f92f6af356293e4f4bec202bb/romancal/pipeline/exposure_pipeline.py#L105-L109
which means that `meta.filename` is not overwritten here:
https://github.com/spacetelescope/romancal/blob/043776069a21ee6f92f6af356293e4f4bec202bb/romancal/pipeline/exposure_pipeline.py#L117-L118
and the `meta.filename` in the input file does not match the filename (r0000101001001001001_01101_0001_WFI01_ALL_SATURATED_uncal.asdf):
```
In [10]: af['roman']['meta']['filename']
Out[10]: 'r0000101001001001001_01101_0001_WFI01_uncal.asdf'
```
making the output filename incorrect.

**Comment by ddavis-stsci** [comment]
I'll add this to the procedure. It should not affect anything for the 
reg tests. The saturated file is created by hand
and I'll add that to the procedure. This has been in place since May 30 
and has not caused problems with the regression
testing, see https://plwishmaster.stsci.edu:8081/job/RT/job/romancal/1008/

if fact the filename should be ignored with this setting

ignore_asdf_paths = {'ignore': ['meta.[date, filename]

**Comment by braingram** [comment]
Thanks!

So it sounds like this is a new issue introduced by #802 and not an issue with the truth file?

To summarize, prior to #802 `meta.filename` was overwritten during exposure pipeline, which allowed `test_processing_pipeline_all_saturated` to generate a file with name `r0000101001001001001_01101_0001_WFI01_ALL_SATURATED_cal.asdf` matching the one expected by the regression test. After #802, `meta.filename` is not overwritten so the same test produces a file `r0000101001001001001_01101_0001_WFI01_cal.asdf` which leads to the `FileNotFoundError` when the regression tests attempts to open a file that doesn't exist.

**Comment by ddavis-stsci** [comment]
I've fixed the input filename and uploaded it to artifactory. I hope 
this error will go away.
uncal = 
rdm.open('r0000101001001001001_01101_0001_WFI01_ALL_SATURATED_uncal.asdf')
 >>> uncal.meta.filename
'r0000101001001001001_01101_0001_WFI01_ALL_SATURATED_uncal.asdf'

**Comment by braingram** [comment]
Thanks for the update.
I ran just this regtest on jenkins and it's now showing a different error:
https://plwishmaster.stsci.edu:8081/job/RT/job/Roman-devdeps/364/testReport/romancal.regtest/test_wfi_pipeline/_unstable_deps__test_processing_pipeline_all_saturated/
```
>       assert model.meta.cal_step.linearity == "SKIPPED"
E       AssertionError: assert 'COMPLETE' == 'SKIPPED'
E         - SKIPPED
E         + COMPLETE

/srv/jenkins/workspace/RT/Roman-devdeps/clone/romancal/regtest/test_wfi_pipeline.py:455: AssertionError
```

Is preferring 'meta.filename' over the actual filename preferred? If so, does this match what happens when files are read from an association? Looking at the code it looks like for those files the filename would take precedence:
https://github.com/spacetelescope/romancal/blob/043776069a21ee6f92f6af356293e4f4bec202bb/romancal/pipeline/exposure_pipeline.py#L105-L118

**Comment by ddavis-stsci** [comment]
OK, the truth file is updated as well. Note: all of these files will be 
updated once INS has new input files for build 11

I'll have to look more closely at the filename precedence. Part of the 
logic is how stpipe handles the filename
and generates the output filename. That makes things more complicated.

**Comment by ddavis-stsci** [comment]
Brett you might want to close this issue. This is fixed in [rcal-631](https://jira.stsci.edu/projects/RCAL/issues/RCAL-631)

**Comment by braingram** [comment]
Thanks for pinging me on this.
I just tested main with `r0000101001001001001_01101_0001_WFI01_uncal.asdf` by running:
```
strun romancal.pipeline.ExposurePipeline r0000101001001001001_01101_0001_WFI01_uncal.asdf
```
This produces a file `r0000101001001001001_01101_0001_WFI01_cal.asdf`.

However, if I copy `r0000101001001001001_01101_0001_WFI01_uncal.asdf` to `foo_uncal.asdf` and then run the same command (using the copied file):
```
strun romancal.pipeline.ExposurePipeline foo_uncal.asdf
```
it produces a file with what appears to be an incorrect filename `r0000101001001001001_01101_0001_WFI01_cal.asdf`. With the relevant log message:
```
2023-08-29 15:52:01,477 - stpipe.ExposurePipeline - INFO - Saved model in r0000101001001001001_01101_0001_WFI01_cal.asdf
```
I will update the title of this issue to hopefully a more descriptive title.

**Comment by ddavis-stsci** [comment]
You cannot just change the filename you need to update meta.filename to match. So you need to copy r0000101001001001001_01101_0001_WFI01_uncal.asdf to foo_uncal.asdf
and set the meta.filename = 'foo_uncal.asdf'

This is built into stpipe and is not something romancal controls. It looks like in stpipe you can use the input filename as a template for the output filename,
```
    def default_output_file(self, input_file=None):
        """Create a default filename based on the input name"""
        output_file = input_file

```
if meta.filename is not present. I haven't played with this. 
We need to use meta.filename here for the case that the input file is an association file, 
e.g. r00001-o001_image_001_asn.json
we don't really want the output file to be r00001-o001_image_001_cal.json(??)

**Comment by braingram** [comment]
Thanks for the response.

It appears that the output filename (used by stpipe) is set here in the romancal `ExposurePipeline`:
https://github.com/spacetelescope/romancal/blob/46ce69ea44de0f1686aa547c697f395517a7f965/romancal/pipeline/exposure_pipeline.py#L181

For a non-association run where the filename (string) is provided as an input. Prior to #802 the input filename was stored to `input_filename` here:
https://github.com/spacetelescope/romancal/blob/04895378d50a312093c55a71508356c2f7000c14/romancal/pipeline/exposure_pipeline.py#L69
prior to the `input` being overwritten as the opened model:
https://github.com/spacetelescope/romancal/blob/04895378d50a312093c55a71508356c2f7000c14/romancal/pipeline/exposure_pipeline.py#L74
after the model is opened `input_filename` is assigned to `input.meta.filename` which meant that the above run of `foo_uncal.asdf` would produce a `foo_cal.asdf` file.

On the current main branch (after #802) the line that assigned `input_filename`:
https://github.com/spacetelescope/romancal/blob/46ce69ea44de0f1686aa547c697f395517a7f965/romancal/pipeline/exposure_pipeline.py#L73
has no effect as `input_filename` is overwritten as `None` before it is used:
https://github.com/spacetelescope/romancal/blob/46ce69ea44de0f1686aa547c697f395517a7f965/romancal/pipeline/exposure_pipeline.py#L109
which results in this if conditional failing and `input.meta.filename` not being assigned:
https://github.com/spacetelescope/romancal/blob/46ce69ea44de0f1686aa547c697f395517a7f965/romancal/pipeline/exposure_pipeline.py#L117-L118

So to summarize, prior to #802 the actual filename was used, after #802 `meta.filename` was used. This resulted in the regression test failure that caused this issue to be opened. Looking at the code, I suspect that for files in an association the actual filename will be used and not `meta.filename` of the file in the association but I do not have an association to test this.

I will follow this up with a PR that should hopefully help clarify this issue.

**Comment by braingram** [comment]
This is no longer an issue now that `meta.filename` is updated on read.
